compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -53,42 +57,60 @@ static SmallVector<ReassociationIndices>
 getCollapsibleLoops(linalg::GenericOp genericOp) {
   SmallVector<ReassociationIndices> contiguousLoops;
 
-  SmallVector<unsigned> pDims;
+  SmallVector<unsigned> pDims, rDims;
   genericOp.getParallelDims(pDims);
-  if (pDims.size() < 2)
-    return contiguousLoops;
-
-  llvm::SmallDenseSet<unsigned> pLoops(pDims.begin(), pDims.end());
+  genericOp.getReductionDims(rDims);
+  llvm::SmallDenseSet<unsigned> pDimsSet, rDimsSet;
+  pDimsSet.insert(pDims.begin(), pDims.end());
+  rDimsSet.insert(rDims.begin(), rDims.end());
 
   auto hasAllMapsSameSequence = [&](AffineExpr preExpr, AffineExpr nextExpr) {
+    // Either both `preExpr` and `nextExpr` are contiguous, or are missing in
+    // all the maps.
     for (AffineMap map : genericOp.getIndexingMapsArray()) {
-      bool foundSeq = false;
+      // If map has no results, no need to check.
+      if (map.getNumResults() == 0) {
+        continue;
+      }
       for (auto [index, resultExpr] : llvm::enumerate(map.getResults())) {
-        if (resultExpr == nextExpr) {
-          foundSeq = (index > 0 && preExpr == map.getResult(index - 1));
-          break;
+        // If we find the preExpr, we should find the nextExpr.
+        if (resultExpr == preExpr) {
```

Comment:
The `nextExpr` does not have to be the next dim. It could be that the sequence `d0, d2`. Basically the loop ordering of the iteration space is mostly meaningless in Linalg ops. So this should capture more cases.

Reply:
`getResultPosition()` looks like it gives the map result index of the given AffineDimExpr, and this loop looks like it's trying to find the first resultExpr that matches `preExpr`, then checks that the map result at `index + 1` is `nextExpr`.

i.e. it's looking for the existence of `..., preExpr, nextExpr, ...` within the affine map results. Is my understandnig wrong here?